homeassistant/components/frank_energie/config_flow.py

```diff
@@ -0,0 +1,112 @@
+"""Config flow for Picnic integration."""
+from __future__ import annotations
+
+from collections.abc import Mapping
+import logging
+from typing import Any
+
+from python_frank_energie import FrankEnergie
+from python_frank_energie.exceptions import AuthException
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_AUTHENTICATION, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import CONF_AUTH_TOKEN, CONF_REFRESH_TOKEN, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle the config flow for Frank Energie."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._reauth_entry: ConfigEntry | None = None
+
+    async def async_step_login(self, user_input=None, errors=None) -> FlowResult:
+        """Handle login with credentials by user."""
+        if not user_input:
+            username = (
+                self._reauth_entry.data[CONF_USERNAME] if self._reauth_entry else None
+            )
+
+            data_schema = vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME, default=username): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            )
+
+            return self.async_show_form(
+                step_id="login",
+                data_schema=data_schema,
+                errors=errors,
+            )
+
+        async with FrankEnergie() as api:
+            try:
+                auth = await api.login(
+                    user_input[CONF_USERNAME], user_input[CONF_PASSWORD]
+                )
+            except AuthException as ex:
+                _LOGGER.exception("Error during login", exc_info=ex)
+                return await self.async_step_login(errors={"base": "invalid_auth"})
+
+        data = {
+            CONF_USERNAME: user_input[CONF_USERNAME],
+            CONF_AUTH_TOKEN: auth.authToken,
+            CONF_REFRESH_TOKEN: auth.refreshToken,
+        }
+
+        if self._reauth_entry:
+            self.hass.config_entries.async_update_entry(
+                self._reauth_entry,
+                data=data,
+            )
+
+            self.hass.async_create_task(
+                self.hass.config_entries.async_reload(self._reauth_entry.entry_id)
+            )
+
+            return self.async_abort(reason="reauth_successful")
+
+        await self.async_set_unique_id(user_input[CONF_USERNAME])
+        self._abort_if_unique_id_configured()
+
+        return await self._async_create_entry(data)
+
+    async def async_step_user(self, user_input=None) -> FlowResult:
+        """Handle a flow initiated by the user."""
+        if not user_input:
+            data_schema = vol.Schema(
+                {
+                    vol.Required(CONF_AUTHENTICATION): bool,
```

Comment:
Can you maybe tell a bit more on how the authentication works? When does one need it?

Reply:
User can get publicly available market prices for electricity and gas without signing in. 

When authenticated, the user gets price data personalised (e.g. different 'inkoopvergoeding'). And billing information (invoices and cost until now)

Reply:
Technically is it just retrieving a JWT to authenticate for different API endpoints. Personalised API's are not available without.

Reply:
Right, in other integrations where authentication is optional (Forecast.solar and opensky) we do authentication in an options flow. Let me check with some people if this is a must or that this solution would be good as well.